### PR TITLE
MC-164: phpspec for AttributeProcessor

### PR DESCRIPTION
**Pull request by damien-carcel**
| Q                    | A
| -------------------- | ---
| Bug fix?             | No
| New feature?         | No
| BC breaks?           | No
| Tests pass?          | Yes
| Checkstyle issues?*  |  No
| PMD issues?**        | No
| Changelog updated?   | No
| Fixed tickets        | MC-164 (partially)
| DB schema updated?   | No
| Migration script?    | No
| Doc PR               | No

### CODE REVIEW

File: spec/Pim/Bundle/MagentoConnectorBundle/Processor/AttributeProcessorSpec.php [hunk 1]
```diff
@@ -0,0 +1,379 @@
...
+
+    function it_sets_a_store_view_mapping(
+        $clientParametersRegistry,
+        $normalizerGuesser,
+        $storeViewMappingMerger,
+        AttributeNormalizer $attributeNormalizer,
+        MagentoSoapClientParameters $clientParameters,
+        MappingCollection $collection
+    ) {
+        $clientParametersRegistry->getInstance(
```
Comment (willy-ahva) [inline spec/Pim/Bundle/MagentoConnectorBundle/Processor/AttributeProcessorSpec.php]: Argument::cetera() will be clearer

File: spec/Pim/Bundle/MagentoConnectorBundle/Processor/AttributeProcessorSpec.php [hunk 2]
```diff
@@ -0,0 +1,379 @@
...
+            null,
+            null,
+            "/api/soap/?wsdl",
+            "default",
+            null,
+            null
+        )->willReturn($clientParameters);
+
+        $normalizerGuesser->getAttributeNormalizer($clientParameters)->willReturn($attributeNormalizer);
+        $storeViewMappingMerger->setParameters($clientParameters, Argument::cetera())->shouldBeCalled();
```
Comment (willy-ahva) [inline spec/Pim/Bundle/MagentoConnectorBundle/Processor/AttributeProcessorSpec.php]: I don't think you are testing this method. Remove all the line.

File: spec/Pim/Bundle/MagentoConnectorBundle/Processor/AttributeProcessorSpec.php [hunk 3]
```diff
@@ -0,0 +1,379 @@
...
+    function it_is_configurable()
+    {
+        $this->setDefaultLocale('en_US');
+        $this->setWebsite('http://mywebsite.com');
+
+        $this->getDefaultLocale()->shouldReturn('en_US');
+        $this->getWebsite()->shouldReturn('http://mywebsite.com');
+    }
+
+    function it_sets_a_store_view_mapping(
```
Comment (willy-ahva) [inline spec/Pim/Bundle/MagentoConnectorBundle/Processor/AttributeProcessorSpec.php]: This example is not good. You should check that getStoreViewMapping returns you what you put in setStoreViewMapping. What setStoreViewMapping returns you is not the purpose.

File: spec/Pim/Bundle/MagentoConnectorBundle/Processor/AttributeProcessorSpec.php [hunk 4]
```diff
@@ -0,0 +1,379 @@
...
+    }
+
+    function it_gives_an_attribute_code_mapping_in_json(
+        $attributeMappingMerger,
+        MappingCollection $mappingCollection
+    ) {
+        $attributeMappingMerger->getMapping()->willReturn($mappingCollection);
+        $mappingCollection->toArray()->willReturn(['foo']);
+
+        $this->getAttributeCodeMapping()->shouldReturn('["foo"]');
```
Comment (willy-ahva) [inline spec/Pim/Bundle/MagentoConnectorBundle/Processor/AttributeProcessorSpec.php]: This is not json, so what do test ? Json or return of method ? And should be in it_is_configurable.

File: spec/Pim/Bundle/MagentoConnectorBundle/Processor/AttributeProcessorSpec.php [hunk 5]
```diff
@@ -0,0 +1,379 @@
...
+        $webserviceGuesser,
+        Attribute $attribute,
+        AttributeNormalizer $attributeNormalizer,
+        GroupRepository $groupRepository,
+        MagentoSoapClientParameters $clientParameters,
+        MappingCollection $attributeMapping,
+        MappingCollection $storeViewMapping,
+        Webservice $webservice
+    ) {
+        $clientParametersRegistry->getInstance(
```
Comment (willy-ahva) [inline spec/Pim/Bundle/MagentoConnectorBundle/Processor/AttributeProcessorSpec.php]: You can put Argument::cetera() to be clearer

File: spec/Pim/Bundle/MagentoConnectorBundle/Processor/AttributeProcessorSpec.php [hunk 6]
```diff
@@ -0,0 +1,379 @@
...
+            'magentoStoreViews'        => $storeViewList,
+            'axisAttributes'           => ['configurable_attribute_code'],
+            'create'                   => true,
+        ];
+
+        $attributeNormalizer->normalize($attribute, 'MagentoArray', $context)->willThrow(
+            'Akeneo\Bundle\BatchBundle\Item\InvalidItemException'
+        );
+
+        $this->shouldThrow('Akeneo\Bundle\BatchBundle\Item\InvalidItemException')->during('process', [$attribute]);
```
Comment (willy-ahva) [inline spec/Pim/Bundle/MagentoConnectorBundle/Processor/AttributeProcessorSpec.php]: You should check what are arguments of this exception.

File: spec/Pim/Bundle/MagentoConnectorBundle/Processor/AttributeProcessorSpec.php [hunk 7]
```diff
@@ -0,0 +1,379 @@
...
+                    'help'     => 'pim_magento_connector.export.magentoUrl.help',
+                    'label'    => 'pim_magento_connector.export.magentoUrl.label',
+                ],
+            ],
+            'wsdlUrl' => [
+                'options' => [
+                    'required' => true,
+                    'help'     => 'pim_magento_connector.export.wsdlUrl.help',
+                    'label'    => 'pim_magento_connector.export.wsdlUrl.label',
+                    'data'     => MagentoSoapClientParameters::SOAP_WSDL_URL,
```
Comment (willy-ahva) [inline spec/Pim/Bundle/MagentoConnectorBundle/Processor/AttributeProcessorSpec.php]: You should put the hard coded constant.

File: spec/Pim/Bundle/MagentoConnectorBundle/Processor/AttributeProcessorSpec.php [hunk 8]
```diff
@@ -0,0 +1,379 @@
...
+                    'help'     => 'pim_magento_connector.export.httpPassword.help',
+                    'label'    => 'pim_magento_connector.export.httpPassword.label',
+                ],
+            ],
+            'defaultStoreView' => [
+                'options' => [
+                    'required' => false,
+                    'help'     => 'pim_magento_connector.export.defaultStoreView.help',
+                    'label'    => 'pim_magento_connector.export.defaultStoreView.label',
+                    'data'     => $this->getDefaultStoreView(),
```
Comment (willy-ahva) [inline spec/Pim/Bundle/MagentoConnectorBundle/Processor/AttributeProcessorSpec.php]: nope

**Comment by willy-ahva** [comment]
You check well what the processor returns in it_processes_an_attribute, but you can also check what you send to the normalizer in terms of what you do in your processor. Because modify and send good data to the normalizer is the purpose of this processor I think.

File: spec/Pim/Bundle/MagentoConnectorBundle/Processor/AttributeProcessorSpec.php (hunk 2 above)
Comment (damien-carcel) [inline spec/Pim/Bundle/MagentoConnectorBundle/Processor/AttributeProcessorSpec.php]: Yes I am, the test breaks if I remove it.

File: spec/Pim/Bundle/MagentoConnectorBundle/Processor/AttributeProcessorSpec.php [hunk 9]
```diff
@@ -0,0 +1,445 @@
...
+        $this->getWebsite()->shouldReturn('my_web_site');
+
+        $this->setDefaultLocale('fr_FR');
+        $this->setWebsite('mon_site_internet');
+
+        $this->getDefaultLocale()->shouldReturn('fr_FR');
+        $this->getWebsite()->shouldReturn('mon_site_internet');
+    }
+
+    function it_sets_a_store_view_mapping(
```
Comment (willy-ahva) [inline spec/Pim/Bundle/MagentoConnectorBundle/Processor/AttributeProcessorSpec.php]: If you want to verify if this class is able to able to set a store view mapping, you should set a store view mapping et verify with the getter if it was set correctly.
Here you verify what is the return of your setter, it's not the same purpose.

File: spec/Pim/Bundle/MagentoConnectorBundle/Processor/AttributeProcessorSpec.php [hunk 10]
```diff
@@ -0,0 +1,445 @@
...
+                    'target' => 'en_us',
+                    'deletable' => true,
+                ]
+            ]
+        );
+
+        $this->setStoreviewMapping('{"en_US":{"source":"en_US","target":"en_us"}}')->shouldReturn($this);
+    }
+
+    function it_sets_an_attribute_code_mapping(
```
Comment (willy-ahva) [inline spec/Pim/Bundle/MagentoConnectorBundle/Processor/AttributeProcessorSpec.php]: Same comment here

**Pull request merged**